core/src/main/java/org/apache/brooklyn/core/OsgiLauncher.java

```diff
@@ -0,0 +1,557 @@
+import org.apache.brooklyn.core.mgmt.internal.LocalManagementContext;
+import org.apache.brooklyn.core.mgmt.internal.ManagementContextInternal;
+import org.apache.brooklyn.core.mgmt.persist.BrooklynMementoPersisterToObjectStore;
+import org.apache.brooklyn.core.mgmt.persist.BrooklynPersistenceUtils;
+import org.apache.brooklyn.core.mgmt.persist.PersistMode;
+import org.apache.brooklyn.core.mgmt.persist.PersistenceObjectStore;
+import org.apache.brooklyn.core.mgmt.rebind.PersistenceExceptionHandlerImpl;
+import org.apache.brooklyn.core.mgmt.rebind.RebindManagerImpl;
+import org.apache.brooklyn.core.objs.BrooklynTypes;
+import org.apache.brooklyn.core.server.BrooklynServerConfig;
+import org.apache.brooklyn.core.server.BrooklynServerPaths;
+import org.apache.brooklyn.util.core.ResourceUtils;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.exceptions.FatalConfigurationRuntimeException;
+import org.apache.brooklyn.util.exceptions.FatalRuntimeException;
+import org.apache.brooklyn.util.exceptions.RuntimeInterruptedException;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.io.FileUtil;
+import org.apache.brooklyn.util.os.Os;
+import org.apache.brooklyn.util.stream.Streams;
+import org.apache.brooklyn.util.text.Strings;
+import org.apache.brooklyn.util.time.Duration;
+import org.apache.brooklyn.util.time.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Initializer for brooklyn-core when running in an OSGi environment.
+ *
+ * Temporarily here; should be totally contained in blueprint beans' init-methods.
+ */
+public class OsgiLauncher {
+
+    private static final Logger log = LoggerFactory.getLogger(OsgiLauncher.class);
+
+    private ManagementContext managementContext;
+
+    private boolean ignoreCatalogErrors = true;
+    private boolean ignorePersistenceErrors = true;
+
+    private HighAvailabilityMode highAvailabilityMode = HighAvailabilityMode.DISABLED;
+
+    private PersistMode persistMode = PersistMode.DISABLED;
+    private String persistenceDir = null;
+    private String persistenceLocation = null;
+    private Duration persistPeriod = Duration.ONE_SECOND;
+
+    public void start() {
+        // Configure launcher
+//        ShutdownHandler shutdownHandler = new AppShutdownHandler();
+        try {
+//            initBrooklynProperties();
+            initCatalog();
+
+            // TODO
+//            launcher.stopWhichAppsOnShutdown(stopWhichAppsOnShutdownMode);
+//            launcher.shutdownHandler(shutdownHandler);
+//
+//            computeAndSetApp(launcher, utils, loader);
+//            customize(launcher);
+//        } catch (FatalConfigurationRuntimeException e) {
+//            throw e;
+//        } catch (Exception e) {
+//            throw new FatalConfigurationRuntimeException("Fatal error configuring Brooklyn launch: " + e.getMessage(), e);
+//        }
+//
+//        try {
```

Comment:
* `initBrooklynProperties()` - agree long term, but it's a lot of work to have everything the "karaf way" from the beginning. I'd prefer if we split it in two stages: 1. have Brooklyn usable in Karaf, 2. Make Brooklyn a good OSGi citizen. This will allow us to work on gradually improving the code base instead of all or nothing approach. 
* Regular users won't be interacting directly with karaf, instead through the cli and webconsole. They will need the process started and shut down at some point. We need a way to shut the process down and optionally terminate all apps as part of that so we clean up cloud resources nicely. This is not so much about backwards compatibility but as having the feature available - don't think the launcher is the place to support this in karaf
* `computeAndSetApp` as part of the cli can be TBD later, not needed right away (but good to have a comment that it is missing)
* `customize(launcher)` - downstream projects have their own `Main` + `Launcher` implementations. This is a generic callback to custom code to do any additional initializations specific to the project. Would be nice to have an infrastructure for downstream projects to influence initialization, but not needed immediately.

To summarize:
  * CLI - needs additional discussion how it maps to the Karaf world, not needed immediately. Personally I believe that it would be nice to support Brooklyn's CLI in headless karaf.
  * `brooklyn.properties` - it would be easy to support most of it so I think we should support it as a transition to Karaf config.
  * next steps would be making sure we can deploy BasicEntity
  * deploy MachineEntity using a jclouds location (by using MachineEntity through a catalog item + explicit libraries bundle pointing at `brooklyn-software-base`)
  * catalog initialization (as in loading entities through reflection) I see as a thing that should be addressed later, i.e. non-essential. If we can have a catalog loaded from `default.catalog.bom`  would be nice, depending on the amount of work involved.